specs/agents/tracing-spans-compress.md

```diff
@@ -0,0 +1,211 @@
+[Agent spec home](README.md) > [Handling huge traces](tracing-spans-handling-huge-traces.md) > [Compressing spans](tracing-spans-compress.md)
+
+## Compressing spans
+
+To mitigate the potential flood of spans to a backend,
+agents SHOULD implement the strategies laid out in this section to avoid sending almost identical and very similar spans.
+
+While compressing multiple similar spans into a single composite span can't fully eliminate the collection overhead,
+it can significantly reduce the impact on the following areas,
+with very little loss of information.
+- Agent reporter queue utilization
+- Capturing stack traces, serialization, compression, and sending events to APM Server
+- Potential to re-use span objects, significantly reducing allocations
+- Downstream effects like reducing impact on APM Server, ES storage, and UI performance
+
+
+### Consecutive-Exact-Match compression strategy
+
+One of the biggest sources of excessive data collection are n+1 type queries and repetitive requests to a cache server.
+This strategy detects consecutive spans that hold the same information (except for the duration)
+and creates a single [composite span](#composite-span).
+
+```
+[                              ]
+GET /users
+ [] [] [] [] [] [] [] [] [] []
+ 10x SELECT FROM users
+```
+
+Two spans are considered to be an exact match if they are of the [same kind](#consecutive-same-kind-compression-strategy) and if their span names are equal:
+- `type`
+- `subtype`
+- `destination.service.resource`
+- `name`
+
+### Consecutive-Same-Kind compression strategy
+
+Another pattern that often occurs is a high amount of alternating queries to the same backend.
+Especially if the individual spans are quite fast, recording every single query is likely to not be worth the overhead.
+
+```
+[                              ]
+GET /users
+ [] [] [] [] [] [] [] [] [] []
+ 10x Calls to mysql
+```
+
+Two spans are considered to be of the same type if the following properties are equal:
+- `type`
+- `subtype`
+- `destination.service.resource`
+
+```java
+boolean isSameKind(Span other) {
+    return type == other.type
+        && subtype == other.subtype
+        && destination.service.resource == other.destination.service.resource
+}
+```
+
+When applying this compression strategy, the `span.name` is set to `Calls to $span.destination.service.resource`.
+The rest of the context, such as the `db.statement` will be determined by the first compressed span, which is turned into a composite span.
+
+
+#### Configuration option `same_kind_compression_max_duration`
```

Comment:
Was prefixing this config option with `span_` considered? I know it is already very long. However, other config options somewhat tend to that pattern:

- [`span_stack_trace_min_duration`](https://github.com/elastic/apm/pull/454)
- `transaction_sample_rate`
- `transaction_ignore_urls`
- `transaction_ignore_user_agents`

Reply:
I agree that adding `span_` prefix will make the meaning clearer but the name will become quite long... I'll bring it up at the next agents meeting.

Reply:
+1 for adding the `span_` prefix. Explicit is better than implicit.

Reply:
Done